### DISCUSSION TITLE
How to correctly use ETag with the paginated responses in REST API

### DISCUSSION THREAD

**Discussion by denis-bigbrassband-com** [community/community#24687] [Apps, API and Webhooks]
<p>Hello all,</p>
<p>I’m trying to implement ETag usage, but I was unable to find information on how to correctly use ETag for paginated responses. For example, I send the following request (assuming that the answer has more than 20 entries):</p>
<pre><code class="lang-auto">GET user/repos?per_page=20
</code></pre>
<p>GitHub response 1 is:</p>
<pre><code class="lang-auto">HTTP 200 OK
ETag: W/"abcdef"
Link: xxxx
...
</code></pre>
<p>From the ‘Link’ field I get a request for the next page and send it:</p>
<pre><code class="lang-auto">GET user/repos?per_page=20&amp;page=2
</code></pre>
<p>GitHub response 2 is:</p>
<pre><code class="lang-auto">HTTP 200 OK
ETag: W/"123456"
Link: xxxx
...
</code></pre>
<p>… and so on.</p>
<p>Now I want to use ETag to get the data again, and send the following request:</p>
<pre><code class="lang-auto">GET user/repos?per_page=20
If-None-Match: "abcdef"
</code></pre>
<p>The GitHub response is:</p>
<pre><code class="lang-auto">HTTP 304 Not Modified
</code></pre>
<p>My question is – how I should interpret the GitHub answer:</p>
<ol>
<li>Only the first page has not been modified, and I should make calls for other pages too (with their ETags)? But there is no ‘Link’ header in the response…<br>
or</li>
<li>The entire data array has not been modified including all subsequent pages, so no additional requests are required?</li>
</ol>
<p>Thank you.</p>

**Comment by imjohnbo** [comment, marked as answer]
<p>Hi <a class="mention" href="/denis-bigbrassband-com">@denis-bigbrassband-com</a> :wave:</p>
<p>Thanks for the question. The ETag is returned per page and is only applicable to the resources in that page (so your interpretation <code>1</code>).</p>
<p>In general, I would recommend conditional requests with large page sizes, and webhooks when possible.</p>
<p>Read more here:</p>
<ul>
<li><a href="https://docs.github.com/en/rest/guides/getting-started-with-the-rest-api#conditional-requests" class="inline-onebox" rel="noopener nofollow ugc">Getting started with the REST API - GitHub Docs</a></li>
<li><a href="https://docs.github.com/en/rest/overview/resources-in-the-rest-api#conditional-requests" class="inline-onebox" rel="noopener nofollow ugc">Resources in the REST API - GitHub Docs</a></li>
</ul>

**Comment by denis-bigbrassband-com** [comment]
<p>Hi John,</p>
<p>Thank you for the confirmation.<br>
Yes, we use the maximum possible page size – 100 (as mentioned <a href="https://docs.github.com/en/rest/overview/resources-in-the-rest-api#pagination" rel="noopener nofollow ugc">here</a>).<br>
But for large lists (yes, we have such ones) it anyway gives 10-100 requests, and I had a hope that this may be reduced to just 1 in case if nothing is changed in the requested data.</p>
<p>Also, it seems that we have to cache the next page link too, as it is absent in the 304 response.</p>
<p>Thank you again for the clarification, it really helped us to implement this feature correctly.</p>
